extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java

```diff
@@ -0,0 +1,263 @@
+            source.stopAndWait();
+            queryExecutor.stopAndWait();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * An implementation of {@link QueryChangeLogListener} for the
+     * {@link QueryManager}.
+     * <p>
+     * When notified of a {@link ChangeType} performs one of the following:
+     * <li>{@link ChangeType#CREATE}: Creates a new query using the
+     * {@link QueryExecutor} provided to the {@link QueryManager}</li>
+     * <li>{@link ChangeType#DELETE}: Deletes a running query by stopping the
+     * {@link QueryExecutor} service of the queryID in the event</li>
+     * <li>{@link ChangeType#UPDATE}: If the query is running and the update is
+     * to stop the query, stops the query. Otherwise, if the query is not
+     * running, it is removed.</li>
+     */
+    private class QueryManagerQueryChange implements QueryChangeLogListener {
+        private final String ryaInstanceName;
+
+        public QueryManagerQueryChange(final String ryaInstanceName) {
+            this.ryaInstanceName = requireNonNull(ryaInstanceName);
+        }
+
+        @Override
+        public void notify(final ChangeLogEntry<QueryChange> queryChangeEvent, final Optional<StreamsQuery> newQueryState) {
+            LOG.debug("New query change event.");
+            final QueryChange entry = queryChangeEvent.getEntry();
+
+            lock.lock();
+            try {
+
+                switch (entry.getChangeType()) {
+                    case CREATE:
+                        LOG.debug("Creating query event.");
+                        if (newQueryState.isPresent()) {
+                            runQuery(ryaInstanceName, newQueryState.get());
+                            LOG.info("Created new query: " + newQueryState.get().toString());
+                        }
+                        break;
+                    case DELETE:
+                        LOG.debug("delete query event.");
+                        if (newQueryState.isPresent()) {
+                            stopQuery(newQueryState.get().getQueryId());
+                            LOG.info("Deleted query: " + newQueryState.get().toString());
+                        } else {
+                            LOG.debug("Delete requested a query that does not exist yet.");
+                        }
+                        break;
+                    case UPDATE:
+                        LOG.debug("update query event.");
+                        if (!newQueryState.isPresent()) {
+                            LOG.debug("Query: " + entry.getQueryId() + " does not exist yet, cannot perform update.");
+                        } else {
+                            final StreamsQuery updateQuery = newQueryState.get();
+                            // if the query is currently inactive, and is updated to
+                            // be active, turn on.
+                            if (!updateQuery.isActive() && entry.getIsActive().or(false)) {
```

Comment:
Kind of the same comment here. You don't know the state before the change was applied within this conditional, so you can't know if you need to start it based on this.